File: FindLargestValueEachTreeRow.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

//Using BFS
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        if(root==null) return new ArrayList<>();

        q.add(root);
         
        List<Integer> result = new ArrayList<>(); 
        while(!q.isEmpty()){
            int max = Integer.MIN_VALUE;
            int len = q.size();

            for (int i=0; i<len;i++){
                TreeNode currNode =  q.remove();
                max = Math.max(currNode.val,max);
                if(currNode.left!=null){
                    q.add(currNode.left);

                }
                if(currNode.right!=null){
                    q.add(currNode.right);

                }
            }
            result.add(max);
        }
        return result;
        
    }
}

File: FindLargestValueEachTreeRow.py
# need to level order traversal BFS
# queue with index of root 
# let first element of q be row_max , iterate over to find maximum
# popleft and append children , append level row max
# time complexity 0(n) [visiting each node once] space complexity 0(n) for the queue assuming
# it could be unbalanced


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root :
            return []

       
        q = deque([root])
        
        res = []
        while q :
            row_max = q[0].val
            qLen = len(q)
            for _ in range(qLen):
                node = q.popleft()
                row_max = max(row_max , node.val)
                if node.left :
                    q.append(node.left)
                if node.right :
                    q.append(node.right)
            res.append(row_max)
        return res
           

            